File: solutions/prolog/complex-numbers/1/complex_numbers.pl
real((X,_), X).
imaginary((_, X), X).

conjugate((R,I), (R, II)) :- II is -I.
abs((R, I), RR) :- RR is sqrt(R * R + I * I).

add((A, B), (C, D), (R, I)) :- R is A + C, I is B + D.
sub((A, B), (C, D), (R, I)) :- R is A - C, I is B - D.

% (a + bi) * (c + di) = (a * c) + (a * di) + (bi * c) + (bi * di)
% (a + bi) * (c + di) = (a * c - b * d) + (a * d + b * c) * i
mul((A, B), (C, D), (R, I)) :- R is A * C - B * D, I is A * D + B * C.
% (a + bi) / (c + di) = (a + bi) * (c - di) / (c + di) * (c - di)
% = (a * c + b * d) + (b * c - a * d)i / (c^2 + d^2)
div((A, B), (C, D), (R, I)) :-
    R is (A * C + B * D) / (C * C + D * D),
    I is (B * C - A * D) / (C * C + D * D).

File: solutions/clojure/complex-numbers/1/src/complex_numbers.clj
(ns complex-numbers)


(defn real [[a b]]
  a)


(defn imaginary [[a b]]
  b)


(defn abs [[a b]]
  (Math/sqrt (+ (* a a) (* b b))))



(defn conjugate [[a b]]
  [a (- b)])


(defn add [[a b] [c d]]
  [(+ a c) (+ b d)])


(defn sub [[a b] [c d]]
  [(- a c) (- b d)])


(defn mul [[a b] [c d]]
  [(- (* a c) (* b d)) (+ (* a d) (* b c))])



(defn div [a b]
  (let [[x y] (mul a (conjugate b))
        [r _] (mul b (conjugate b))]
    [(/ x r 1.0) (/ y r 1.0)]))
